### PR TITLE
Refactoring to PSRs and middlewares

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | yes
| Tests pass?   | not yet
| Fixed issues  | #11 

- Web application doesn't depend on core application anymore (there's no sense in trying to unify web/console).
- PSR-7 request/response instead of our own.
- Emitting response separated, interface introduced with SAPI implementation. Other implementations could be done for non-dying web servers such as [RoadRunner](https://github.com/spiral/roadrunner).
- PSR middlewares.
- Routing extracted into [yiisoft/router](https://github.com/yiisoft/router).

### CODE REVIEW

File: src/Application.php [hunk 1]
```diff
@@ -7,139 +7,35 @@
...
-    /**
-     * @param string $value the homepage URL
-     */
-    public function setHomeUrl($value)
-    {
-        $this->_homeUrl = $value;
-    }
-
-    public function getAssetManager()
+    public function __construct(ContainerInterface $container)
```
Comment (hiqsol) [inline src/Application.php]: why not:

```php
public function __construct(
	RequestFactoryInterface $requestFactory,
	DispatcherInterface $dispatcher,
	EmitterInterface $emitter
);
```

Comment (samdark) [reply src/Application.php]: Yes. Could be done.

**Comment by samdark** [comment]
Gave up on implementing our own routing for now. Made it an interface + adapter (some code taken from Zend Expressive Router). The adapter available now is for FastRoute.

The application itself looks like the following:

```php
require '../yii-web/vendor/autoload.php';

$config = require __DIR__ . '/container.php';

$container = new \yii\di\Container($config);

(new \yii\web\Application(
    $container->get(\yii\web\ServerRequestFactory::class),
    $container->get(\yii\web\MiddlewareDispatcher::class),
    $container->get(\yii\web\emitter\EmitterInterface::class)
))->run();
```

And the container.php:

```php
use FastRoute\RouteCollector;
use Psr\Container\ContainerInterface;
use Psr\Http\Message\RequestFactoryInterface;
use Psr\Http\Message\ResponseFactoryInterface;
use Psr\Http\Message\ServerRequestFactoryInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Message\StreamFactoryInterface;
use Psr\Http\Message\UploadedFileFactoryInterface;
use Psr\Http\Message\UriFactoryInterface;
use yii\di\Reference;
use yii\web\emitter\EmitterInterface;
use yii\web\emitter\SapiEmitter;
use yii\web\MiddlewareDispatcher;
use yii\web\router\adapters\FastRoute;
use yii\web\router\Route;
use yii\web\router\RouterInterface;
use yii\web\router\UrlGeneratorInterface;
use yii\web\router\UrlMatcherInterface;

return [
    // PSR-17 factories:
    RequestFactoryInterface::class => \Nyholm\Psr7\Factory\Psr17Factory::class,
    ServerRequestFactoryInterface::class => \Nyholm\Psr7\Factory\Psr17Factory::class,
    ResponseFactoryInterface::class => \Nyholm\Psr7\Factory\Psr17Factory::class,
    StreamFactoryInterface::class => \Nyholm\Psr7\Factory\Psr17Factory::class,
    UriFactoryInterface::class => \Nyholm\Psr7\Factory\Psr17Factory::class,
    UploadedFileFactoryInterface::class => \Nyholm\Psr7\Factory\Psr17Factory::class,

    // custom stuff

    EmitterInterface::class => SapiEmitter::class,

    RouterInterface::class => function (ContainerInterface $container) {
        $collector = new RouteCollector(
            new \FastRoute\RouteParser\Std(),
            new \FastRoute\DataGenerator\GroupCountBased()
        );

        $router = new FastRoute(
            $collector,
            function ($data) {
                return new \FastRoute\Dispatcher\GroupCountBased($data);
            }
        );

        $router->addRoute(Route::get('/')->to(function (ServerRequestInterface $request, \Psr\Http\Server\RequestHandlerInterface $next) {
            $response = new \Nyholm\Psr7\Response();
            $response->getBody()->write('You are at homepage.');
            return $response;
        }));

        $router->addRoute(Route::get('/test/{id:\w+}')->to(function (ServerRequestInterface $request, \Psr\Http\Server\RequestHandlerInterface $next) {
            $bla = $request->getAttribute('id');

            $response = new \Nyholm\Psr7\Response();
            $response->getBody()->write('You are at test with param ' . $bla);
            return $response;
        }));

        return $router;
    },

    UrlMatcherInterface::class => Reference::to(RouterInterface::class),
    UrlGeneratorInterface::class => Reference::to(RouterInterface::class),

    MiddlewareDispatcher::class => function (ContainerInterface $container) {
        $responseFactory = $container->get(ResponseFactoryInterface::class);
        $urlMatcher = $container->get(UrlMatcherInterface::class);

        return new MiddlewareDispatcher([
            new \yii\web\middleware\Router($urlMatcher),
            new \yii\web\middleware\RoutingDispatcher(),
        ], $responseFactory);
    }
];
```

**Comment by samdark** [comment]
The config looks better if we extract closures into factories:

```php
use Nyholm\Psr7\Factory\Psr17Factory;
use Psr\Http\Message\RequestFactoryInterface;
use Psr\Http\Message\ResponseFactoryInterface;
use Psr\Http\Message\ServerRequestFactoryInterface;
use Psr\Http\Message\StreamFactoryInterface;
use Psr\Http\Message\UploadedFileFactoryInterface;
use Psr\Http\Message\UriFactoryInterface;
use yii\di\Reference;
use yii\web\emitter\EmitterInterface;
use yii\web\emitter\SapiEmitter;
use yii\web\MiddlewareDispatcher;
use Yiisoft\Router\RouterInterface;
use Yiisoft\Router\UrlGeneratorInterface;
use Yiisoft\Router\UrlMatcherInterface;
use Yiisoft\Sandbox\FastRouteFactory;
use Yiisoft\Sandbox\MiddlewareDispatcherFactory;

return [
    // PSR-17 factories:
    RequestFactoryInterface::class => Psr17Factory::class,
    ServerRequestFactoryInterface::class => Psr17Factory::class,
    ResponseFactoryInterface::class => Psr17Factory::class,
    StreamFactoryInterface::class => Psr17Factory::class,
    UriFactoryInterface::class => Psr17Factory::class,
    UploadedFileFactoryInterface::class => Psr17Factory::class,

    // custom stuff

    EmitterInterface::class => SapiEmitter::class,
    RouterInterface::class => new FastRouteFactory(),
    UrlMatcherInterface::class => Reference::to(RouterInterface::class),
    UrlGeneratorInterface::class => Reference::to(RouterInterface::class),
    MiddlewareDispatcher::class => new MiddlewareDispatcherFactory()
];
```

FastRouteFactory:

```php
use FastRoute\DataGenerator\GroupCountBased;
use FastRoute\RouteCollector;
use FastRoute\RouteParser\Std;
use Psr\Container\ContainerInterface;
use Yiisoft\Router\FastRoute\FastRoute;

class FastRouteFactory
{
    public function __invoke(ContainerInterface $container)
    {
        $collector = new RouteCollector(
            new Std(),
            new GroupCountBased()
        );

        $router = new FastRoute(
            $collector,
            function ($data) {
                return new \FastRoute\Dispatcher\GroupCountBased($data);
            }
        );

        return $router;
    }
}
```

MiddlewareDispatcherFactory:

```php
use Psr\Http\Message\ResponseFactoryInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\RequestHandlerInterface;
use yii\di\Container;
use yii\web\MiddlewareDispatcher;
use Yiisoft\Router\Middleware\Router;
use Yiisoft\Router\Route;
use Yiisoft\Router\RouterInterface;

class MiddlewareDispatcherFactory
{
    public function __invoke(Container $container)
    {
        /* @var ResponseFactoryInterface $responseFactory */
        $responseFactory = $container->get(ResponseFactoryInterface::class);

        /* @var Router $router */
        $router = $container->get(RouterInterface::class);

        $router->addRoute(Route::get('/')->to(function (ServerRequestInterface $request, RequestHandlerInterface $next) use ($responseFactory) {
            $response = $responseFactory->createResponse();
            $response->getBody()->write('You are at homepage.');
            return $response;
        }));

        $router->addRoute(Route::get('/test/{id:\w+}')->to(function (ServerRequestInterface $request, RequestHandlerInterface $next) use ($responseFactory) {
            $id = $request->getAttribute('id');

            $response = $responseFactory->createResponse();
            $response->getBody()->write('You are at test with param ' . $id);
            return $response;
        }));

        return new MiddlewareDispatcher([
            new Router($router),
        ], $responseFactory);
    }
}
```

**Pull request merged**